File: lib/app/api/models/chat_model.dart
import 'package:ventes/app/api/models/business_partner_model.dart';
import 'package:ventes/app/api/models/files_model.dart';
import 'package:ventes/app/api/models/user_model.dart';
import 'package:ventes/core/api/model.dart';

class Chat extends Model {
  int? chatid;
  int? chatbpid;
  String? chatmessage;
  String? chatrefname;
  int? chatrefid;
  String? chatreadat;
  int? chatreceiverid;
  BusinessPartner? chatbp;
  User? chatreceiver;
  User? createdbyuser;
  Files? chatfile;

  Chat({
    this.chatid,
    this.chatbpid,
    this.chatmessage,
    this.chatrefname,
    this.chatrefid,
    this.chatfile,
    this.chatreadat,
    this.chatreceiverid,
    this.chatbp,
    this.chatreceiver,
    this.createdbyuser,
    String? createddate,
    String? updateddate,
    int? createdby,
    int? updatedby,
    bool? isactive,
  }) : super(
          createddate: createddate,
          createdby: createdby,
          updateddate: updateddate,
          updatedby: updatedby,
          isactive: isactive,
        );

  Chat.fromJson(Map<String, dynamic> json) {
    chatid = json['chatid'];
    chatbpid = json['chatbpid'];
    chatmessage = json['chatmessage'];
    chatrefname = json['chatrefname'];
    chatrefid = json['chatrefid'];
    chatreadat = json['chatreadat'];
    chatreceiverid = json['chatreceiverid'];

    if (json['chatbp'] != null) chatbp = BusinessPartner.fromJson(json['chatbp']);
    if (json['chatreceiver'] != null) chatreceiver = User.fromJson(json['chatreceiver']);
    if (json['createdbyuser'] != null) createdbyuser = User.fromJson(json['createdbyuser']);
    if (json['chatfile'] != null) chatfile = Files.fromJson(json['chatfile']);
    super.fromJson(json);
  }

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['chatid'] = chatid;
    data['chatbpid'] = chatbpid;
    data['chatmessage'] = chatmessage;
    data['chatrefname'] = chatrefname;
    data['chatrefid'] = chatrefid;
    data['chatfile'] = chatfile;
    data['chatreadat'] = chatreadat;
    data['chatreceiverid'] = chatreceiverid;
    data['chatbp'] = chatbp;
    data['chatreceiver'] = chatreceiver;

    if (chatbp != null) {
      data['chatbp'] = chatbp?.toJson();
    }

    if (chatreceiver != null) {
      data['chatreceiver'] = chatreceiver?.toJson();
    }

    if (createdbyuser != null) {
      data['createdbyuser'] = createdbyuser?.toJson();
    }

    if (chatfile != null) {
      data['chatfile'] = chatfile?.toJson();
    }
    return data;
  }
}

File: server/app/models/chatModel.js
const Model = require('./model.js');
class ChatModel extends Model {
   chatid;
   chatbpid;
   chatmessage;
   chatrefname;
   chatrefid;
   chatfile;
   chatreadat;
   chatreceiverid;
   chatbp;
   chatreceiver;
   createdbyuser;

   constructor(
      chatid,
      chatbpid,
      chatmessage,
      chatrefname,
      chatrefid,
      chatfile,
      chatreadat,
      chatreceiverid,
      chatbp,
      chatreceiver,
      createdbyuser,
      createddate,
      updateddate,
      createdby,
      updatedby,
      isactive,
   ) {
      super(
         createddate,
         createdby,
         updateddate,
         updatedby,
         isactive
      );
      this.chatid = chatid;
      this.chatbpid = chatbpid;
      this.chatmessage = chatmessage;
      this.chatrefname = chatrefname;
      this.chatrefid = chatrefid;
      this.chatfile = chatfile;
      this.chatreadat = chatreadat;
      this.chatreceiverid = chatreceiverid;
      this.chatbp = chatbp;
      this.chatreceiver = chatreceiver;
      this.createdbyuser = createdbyuser;
   };

   static fromJson(json) {
      super.fromJson(json);
      this.chatid = json.chatid;
      this.chatbpid = json.chatbpid;
      this.chatmessage = json.chatmessage;
      this.chatrefname = json.chatrefname;
      this.chatrefid = json.chatrefid;
      this.chatfile = json.chatfile;
      this.chatreadat = json.chatreadat;
      this.chatreceiverid = json.chatreceiverid;
      this.chatbp = json.chatbp;
      this.chatreceiver = json.chatreceiver;
      this.createdbyuser = json.createdbyuser;
      return this;
   }

   toJson() {
      let data = super.toJson();
      data.chatid = this.chatid;
      data.chatbpid = this.chatbpid;
      data.chatmessage = this.chatmessage;
      data.chatrefname = this.chatrefname;
      data.chatrefid = this.chatrefid;
      data.chatfile = this.chatfile;
      data.chatreadat = this.chatreadat;
      data.chatreceiverid = this.chatreceiverid;
      data.chatbp = this.chatbp;
      data.chatreceiver = this.chatreceiver;
      data.createdbyuser = this.createdbyuser;
      return data;
   }
}

module.exports = ChatModel;